sdks/java/io/hadoop-input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/unit/tests/HadoopInputFormatIOTest.java

```diff
@@ -0,0 +1,776 @@
+        .withConfiguration(serConf.getHadoopConfiguration())
+        .withKeyTranslation(myKeyTranslateWithWrongInputType);
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(
+        String.format(HadoopInputFormatIOContants.WRONG_KEY_TRANSLATIONFUNC_ERROR_MSG,
+            serConf.getHadoopConfiguration()
+                .getClass(HadoopInputFormatIOContants.INPUTFORMAT_CLASSNAME, InputFormat.class),
+            serConf.getHadoopConfiguration().getClass(HadoopInputFormatIOContants.KEY_CLASS,
+                Object.class)));
+    read.validate(input);
+
+  }
+
+  /**
+   * This test validates functionality of Read.validate() function when myValueTranslate's (simple
+   * function provided by user for value translation) input type is not same as Hadoop InputFormat's
+   * valueClass(Which is property set in configuration as "value.class").
+   */
+  @Test
+  public void testReadValidationFailsWithWrongInputTypeValueTranslationFunction() {
+    SimpleFunction<LongWritable, String> myValueTranslateWithWrongInputType =
+        new SimpleFunction<LongWritable, String>() {
+          private static final long serialVersionUID = 1L;
+          @Override
+          public String apply(LongWritable input) {
+            return input.toString();
+          }
+        };
+    HadoopInputFormatIO.Read<Text, String> read = HadoopInputFormatIO.<Text, String>read()
+            .withConfiguration(serConf.getHadoopConfiguration())
+            .withValueTranslation(myValueTranslateWithWrongInputType);
+    String expectedMessage =
+        String.format(HadoopInputFormatIOContants.WRONG_VALUE_TRANSLATIONFUNC_ERROR_MSG,
+            serConf.getHadoopConfiguration()
+                .getClass(HadoopInputFormatIOContants.INPUTFORMAT_CLASSNAME, InputFormat.class),
+            serConf.getHadoopConfiguration()
+                .getClass(HadoopInputFormatIOContants.VALUE_CLASS, Object.class));
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(expectedMessage);
+    read.validate(input);
+  }
+
+  @Test
+  public void testReadingData() throws Exception {
+    HadoopInputFormatIO.Read<Text, Employee> read = HadoopInputFormatIO.<Text, Employee>read()
+        .withConfiguration(serConf.getHadoopConfiguration());
+    List<KV<Text, Employee>> expected = UnitTestUtils.getEmployeeData();
+    PCollection<KV<Text, Employee>> actual = p.apply("ReadTest", read);
+    PAssert.that(actual).containsInAnyOrder(expected);
+    p.run();
+  }
+
+  /**
+   * This test validates behavior of HadoopInputFormatSource if RecordReader object creation fails
+   * in start() method.
+   */
+  @Test
+  public void testReadersStartIfCreateRecordReaderFails() throws Exception {
+    SerializableConfiguration serConf = 
+      loadTestConfiguration(
+            BadCreateReaderInputFormat.class, 
+            Text.class, 
+            Employee.class);
+    HadoopInputFormatBoundedSource<Text, Employee> parentHIFSource =
```

Comment:
Many of these reader tests seem to follow a pattern with creating a config, the source, running split and then using the values from the split. Does it make sense to extract out the first couple steps?